### DISCUSSION TITLE
distinct colors for neighboring points?

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#63] [General]
I saw this question in fosstodon: 'I want to distinguish a few hundred data points in. Some of them are adjacent to each other. I’d like these to have contrasting colors or different patterns to easily distinguish them, but I don’t need to connect these to some underlying value, so reusing colors or patterns is not a problem, as long as bordering data points are distinct. Any recommendations on how to do this?'

I thought it is an interesting question and do not know an answer though it seemed like some kind of voronoi partitioning and then applying colors using some kind of four/five color theorem algorithm might be a way forward.  I see gg-choropleths a ton, but not so much using colors to distinguish areas...

**Comment by yjunechoe** [comment]
I saw that post too but I failed to understand the question 😅 - I hope they share the final plot when they settle on a solution!

**Reply by EvaMaeRey** [reply]
I think this is the question, probably partially satisfing @teunbrand's stack overflow tendencies but probably not totally. 

tibble(x = runif(300),
           y = runif(300)) |>
mutate(distinct_colors_for_neighbors = alogrithm(x,y, max_colors = 5)) |>
ggplot() + 
aes(x = x, y = y) + 
geom_point() + 
aes(color = I(distinct_colors_for_neighbors))

I think the question is:   is there an algorithm that does this directly or an indirect way.

**Comment by teunbrand** [comment]
I hate to invoke my stackoverflow-posting self, but can you give a minimal reproducible example of some plot you'd like to apply this to? Ideally with only a few datapoints so that we can conscientiously apply the rules.

**Reply by EvaMaeRey** [reply]
For the polygons computed w/ voronoi tile, fill with a color such that no neighboring tiles have the same color - https://en.wikipedia.org/wiki/Five_color_theorem - and then make that the point color, was my idea.   I just actually don't think I've seen five color theorem every used w/ ggplot2; 

``` r
library(tidyverse)
tibble(x = runif(50),
       y = runif(50)) |>
  ggplot() +
  aes(x = x, y = y) +
  ggforce::geom_voronoi_tile() +
  geom_point()
```

![](https://i.imgur.com/bt4Cr5O.png)

``` r

layer_data(i = 1) |> head()
#>   group        x         y     theta    orig_x    orig_y PANEL colour      fill
#> 1  1:-1 0.588331  0.163973 0.2763248 0.4864159 0.1350717     1     NA #333333FF
#> 2  1:-1 0.552576  0.201962 0.7908848 0.4864159 0.1350717     1     NA #333333FF
#> 3  1:-1 0.487244  0.234176 1.5624417 0.4864159 0.1350717     1     NA #333333FF
#> 4  1:-1 0.424731  0.084713 3.8262477 0.4864159 0.1350717     1     NA #333333FF
#> 5  1:-1 0.514110 -0.061979 4.8520168 0.4864159 0.1350717     1     NA #333333FF
#> 6  2:-1 0.765740  0.742261 2.1449715 0.8054732 0.6808376     1     NA #333333FF
#>   linewidth linetype alpha
#> 1       0.5        1    NA
#> 2       0.5        1    NA
#> 3       0.5        1    NA
#> 4       0.5        1    NA
#> 5       0.5        1    NA
#> 6       0.5        1    NA
```

<sup>Created on 2024-11-20 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by mjskay** [reply]
I'm not sure I understand the motivation from this example --- if the points are not overlapping, they are already easy to distinguish; if the points are overlapping, it's unclear to me that color (and not something like opacity) would be the appropriate solution.

Or is the problem that there are *groups* of points that need to be distinguished from each other?

**Reply by EvaMaeRey** [reply]
Yeah.  I'm not sure I'm representing the question right, and even if so, if instinct is the most useful.  ggblend definitely came to mind looking at this question -- feels like at least a similar problem space.

**Comment by teunbrand** [comment]
Well here is my naive approach which likely isn't optimal. It just greedily grabs the first colour a point doesn't see in its neighbours. No guarantees it sticks to <=5 colours. This was fun, but I won't be further nerdsniped  today

``` r
library(tidyverse)
library(igraph)

# Init data
df <- tibble(x = runif(50), y = runif(50))

# Make a graph of delaunay triangulation of the points
delaunay <- deldir::deldir(df$x, df$y)$delsgs
delaunay <- cbind(delaunay$ind1, delaunay$ind2)
g <- igraph::graph_from_edgelist(delaunay, directed = FALSE)

# For every vertex, search an unused colour
for (v in seq_along(V(g))) {
  v <- V(g)[v] # Index to current vertex
  nn <- ego(g, nodes = v)[[1]] # Current vertex to neighbours
  
  # Determine which colours are already in use
  nn_colours <- vertex_attr(g, "colour", index = nn) %||% 0
  nn_colours <- unique(nn_colours[!is.na(nn_colours)])
  if (length(nn_colours) == 0) {
    nn_colours <- 0
  }
  mx <- max(nn_colours)
  
  if (mx == 0) { # When no colours have been assigned to neighbours
    i <- 1L
  } else {
    # Try to find lowest colour that isn't in use by neighbours
    i <- setdiff(seq_len(mx), nn_colours)
    if (length(i) > 0) {
      i <- i[1]
    } else {
      i <- mx + 1 # If not found, increase local maximum
    }
  }
  g <- set_vertex_attr(g, "colour", index = v, value = i)
}

df$colour <- vertex_attr(g, "colour")

ggplot(df) +
  aes(x = x, y = y) +
  ggforce::geom_voronoi_tile(aes(fill = factor(colour), group = 1)) +
  geom_point()
```

![](https://i.imgur.com/gMKnUMm.png)

<sup>Created on 2024-11-20 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
Yes, you indeed just saw a unicorn: one of the legit reasons to use a for-loop is if one iteration depends on the previous one 🦄

**Reply by EvaMaeRey** [reply]
Very nice!  I also need to think about this no longer today!  sorry for the distraction. Seems like it should be pretty easy to link up the colour designations back to the points - but not today!

**Reply by EvaMaeRey** [reply]
actually it's all done isn't it!?  And colors look spaced out. 🚀  

ggplot(df) +
  aes(x = x, y = y, color = factor(colour)) +
  geom_point() + 
  scale_color_viridis_d()

![image](https://github.com/user-attachments/assets/77ca33b1-4852-441a-818c-0b4c0c9133ad)

**Reply by teunbrand** [reply]
Well yeah it fills the requirement that adjacent voronoi cells don't have the same colours.
But it isn't optimal in that it doesn't adhere to the 4 or 5 colour theorem: I've set n = 5000 simulated points and it needed 7 colours to cover it.

**Reply by EvaMaeRey** [reply]
Ah.  Yeah I saw 6 even with 50.  higher number colors are noticeably less used too, which I guess isn't ideal.  I guess there is probably a nice theorem implementation already out there that can be pulled in.  I'd be interested in it for geographic maps too.  There's this https://github.com/hunzikp/MapColoring ... it's working in the sp world instead of sf so I expect it might be slightly harder to connect up into a workflow.  but maybe not.

**Reply by teunbrand** [reply]
I thought the whole sp infra was deprecated not too long ago.  In any case, my strategy wouldn't work with geographic data. Or at least, you'd have to replace the delaunay triangulation graph with a 'shared border graph' (if that is a thing). This can become quite tricky as there are weird borders. For example: even though the Netherlands and France do not share a border in Europe (as they are separated by Belgium), they do share a border on the Caribean island of Sint Maarten/Saint-Marten.

**Reply by EvaMaeRey** [reply]
I hadn't thought about Sint Maarten!  😞😞😞

**Comment by EvaMaeRey** [comment]
Recently follow up from @loreabad6 also consulting with @Nowosad to checkout tmap::map_coloring!   https://r-tmap.github.io/tmaptools/reference/map_coloring.html, post voronoi

https://fosstodon.org/@loreabad6/114025371594322561  Thanks for the tip!

**Reply by teunbrand** [reply]
Ah sweet, seems you *can* get a shared border graph from an {sf} object! The algorithm seems to do a similar thing but without the {igraph} bits.
